Plutarch/Builtin/List.hs

```diff
@@ -0,0 +1,111 @@
+{-# OPTIONS_GHC -Wno-orphans #-}
+
+module Plutarch.Builtin.List (
+  -- * List type
+  PList (..),
+
+  -- * List construction
+  cons,
+  nil,
+  mkList,
+  singleton,
+  append,
+
+  -- * List access
+  head,
+  contains,
+  atIndex,
+) where
+
+import Data.Proxy
+import Plutarch
+import Plutarch.Bool
+import Plutarch.Builtin
+import Plutarch.Builtin.List.Type
+import Plutarch.Integer (PInteger)
+import Plutarch.Prelude
+import qualified PlutusCore as PLC
+import Prelude hiding (head)
+
+instance ListElemUni (a :: k -> Type) => PlutusType (PList a) where
```

Comment:
If this is the only way of doing this, then we should not do it and instead use Scott-encoded lists in general.

Reply:
We will need to be able to deconstruct built in lists either way. I understand that this machinery is for *building* lists rather than deconstructing them. But, IMO, we will need a way for building builtin lists along the way as well (see: `ListData`).

Unless we are 100% sure no one will ever ask for the ability to create builtin lists - I sincerely suggest we keep a *complete* builtin list implementation with great ergonomics for both building and deconstructing.

Reply:
For deconstructing lists you don't need `ListElemUni`.

Reply:
I understand that - but do we *really* not need to have a way of constructing builtin lists?

Reply:
We can make a fully polymorphic `pmatch` since it's a separate type class now. We can still make a `PCon` instance, but we might still want a separate `pcons` that is fully polymorphic.